[FLINK-6888] [table] Can not determine TypeInformation of ACC type of AggregateFunction when ACC is a Scala class

wuchong opened the pull request:
Currently the `ACC TypeInformation of `org.apache.flink.table.functions.AggregateFunction[T, ACC]` is extracted using `TypeInformation.of(Class)`. When `ACC` is a Scala case class or tuple class, the TypeInformation will fall back to `GenericType` which result in bad performance when state de/serialization. 

This PR fix this issue by extracting ACC TypeInformation when calling `TableEnvironment.registerFunction()`.

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -1395,50 +1399,25 @@ object AggregateUtil {
-      }
+    // create accumulator type information for every aggregate function
+    aggregates.zipWithIndex.foreach { case (agg, index) =>
+      accTypes(index) = getAccumulatorTypeOfAggregateFunction(agg, accTypes(index))
```
I think we need add a check as follows:
```
aggregates.zipWithIndex.foreach { case (agg, index) =>
      if(null ==  accTypes(index)){
        accTypes(index) = getAccumulatorTypeOfAggregateFunction(agg, accTypes(index))
      }
    }
```

wuchong replied:
I don't think so. The `getAccumulatorTypeOfAggregateFunction` can handle the nullable `accTypes(index)`.

sunjincheng121 replied:
If i understand correctly. only build-in AGG need call `getAccumulatorTypeOfAggregateFunction`. 
and all the UDAGG do not need call this method. is that correct? If so, I think if only `null==accTypes(index)` need call this method. Please explain more If I understand incorrectly. thanks!

wuchong replied:
Yes, adding the `null == accTypes(index)` condition can reduce one more time calling `getAccumulatorTypeOfAggregateFunction`.  I will add it.

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/AggSqlFunction.scala:
```diff
@@ -37,24 +36,26 @@ import org.apache.flink.table.functions.utils.UserDefinedFunctionUtils._
     typeFactory: FlinkTypeFactory,
     requiresOver: Boolean)
-  extends SqlUserDefinedAggFunction(
+  extends SqlAggFunction(
```
Sorry, I didn't notice the copied `SqlUserDefinedAggFunction` from Calcite. I did these change because the constructor of Calcite `SqlUserDefinedAggFunction` doesn't match the calling with additional `requiresOver`.

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala:
```diff
@@ -329,6 +337,41 @@ object UserDefinedFunctionUtils {
+    * and uses [[TypeExtractor]] as default return type inference.
+    */
+  def getAccumulatorTypeOfAggregateFunction(
+    aggregateFunction: AggregateFunction[_, _],
```
Good idea. I like it.

wuchong replied:
I find that only the following code block can be extracted. I think maybe we don't need to extract it. What do you think? @sunjincheng121 

```
  val resultType = try {
      val method: Method = aggregateFunction.getClass.getMethod(methodName)
      method.invoke(aggregateFunction).asInstanceOf[TypeInformation[_]]
    } catch {
      case _: NoSuchMethodException => null
      case ite: Throwable => throw new TableException("Unexpected exception:", ite)
    }
```

hustfxj commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala:
```diff
@@ -329,6 +337,41 @@ object UserDefinedFunctionUtils {
+    } else {
+      val accumulator = aggregateFunction.createAccumulator()
+      try {
+        TypeInformation.of(accumulator.getClass)
```
Maybe we use the followed code replace **TypeInformation.of** ?
 ```
val accumulator = TypeExtractor
        .createTypeInfo(aggregateFunction, classOf[AggregateFunction[_, _]],
          aggregateFunction.getClass, 1).asInstanceOf[TypeInformation[_]]
```

wuchong replied:
I think it's a good advice.

StephanEwen commented:
Is the code ever called from Java code, or is that purely within the Scala code base?

Just FYI, the design principles in the DataStream and DataSet API are:
  - Never run `TypeExtractor` for Scala code. Scala code always uses implicits and context bounds for the Type Information
  - Java code entry methods uses the Type Extractor
  - Code that is shared between Java / Scala always gets the Type Information passed in

wuchong commented:
Thanks @StephanEwen .  This PR is aiming to fix the issues you mentioned: it used `TypeExtractor` for Scala code. 

@sunjincheng121  I addressed all your comments. Could you please have a look again ?

sunjincheng121 commented:
@wuchong thanks for the update. It's looks pretty good for me now.
+ to merged.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala:
```diff
@@ -358,20 +358,27 @@ abstract class TableEnvironment(val config: TableConfig) {
 
     // register in SQL API
-    val sqlFunctions = createAggregateSqlFunction(name, function, typeInfo, typeFactory)
+    val sqlFunctions = createAggregateSqlFunction(
```
To be honest, I don't think it is a nice approach to add the acc type to the internal representation of UDAGGs. It is not related to the logic of the function or required for the optimization but rather an internal runtime aspect.

An alternative could be to wrap the functions registered from Scala in a class that extends the `AggregateFunction` interface and holds the original agg function and the acc type information (exposed via the getAccumulatorType()` method). We would need to unwrap it before we translate it. It's not a super nice solution either, but would probably require fewer changes.

What do you think?

wuchong replied:
Hi @fhueske , I think the ACC type extraction is similar to `T` type of `AggreateFunction[T, ACC]` and `T` type of `TableFunction[T]`. Currently, they are both extracted and add to the internal representation of UDAGGs and UDTFs. To keep the consistent, I followed the same way to add the acc type. If we decide to adopt the approach that wrapping the registered functions, it would be better to make the other type extractions follow this approach.

What do you think ?

fhueske replied:
You are right. We use the same approach for the return type `T` of all UDFs. However, the return type is part of the function signature and needs to be known to Calcite for semantic validation. The `ACC` type is only needed for compilation and an engine specific property.  

TBH, I'm not sure which approach is better.

wuchong replied:
@fhueske thanks for your explanation. I tried to wrap the UDAGG in an `AggregateFunctionWrapper`, but find that it is not an easy way. Because I can't override the user-defined contract methods, such as `accumulate`, `retract`, `merge`. And in code generation, we generate the acc type class depend on `createAccumulator` method return type, but the return type of `createAccumulator` of `AggregateFunctionWrapper` can only be `Any` which will result in error.

In addition, I plan to support composite result type for UDAGG. This also need the return type not only to be known to Calcite for semantic validation but also to be known for code generation.

wuchong replied:
Ping @fhueske , what do you think about this?

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/java/BatchTableEnvironment.scala:
```diff
@@ -192,10 +193,14 @@ class BatchTableEnvironment(
+      .getResultTypeOfAggregateFunction(f)
       .asInstanceOf[TypeInformation[T]]
 
+    implicit val accTypeInfo: TypeInformation[ACC] = UserDefinedFunctionUtils
```
I agree with you. But this PR is only to fix Scala classes can't work in the Scala TableEnvironments. In general, users create Scala classes and naturally will use it in Scala TableEnvironments. Registering a Scala class into Java TableEnvironments is not recommended (should forbidden). This is the same with that Scala `T` classes of `Aggregate[T, _] and TableFunction[T]` works in Scala TableEnvironments.

fhueske replied:
UDAGGs are sometimes also shared in libraries. If a UDAGG class is included in a JAR file and loaded into the classpath, a user can hardly tell whether it is implemented in Java or Scala. So, I would not prohibit Scala UDFs in Jave environments. However, functions which are intended to be shared should be implemented with Java types.

wuchong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala:
```diff
@@ -320,12 +348,20 @@ object UserDefinedFunctionUtils {
     } else {
-      TypeExtractor
+      try {
+        TypeExtractor
```
I think it is more safe than `TypeInformation.of(Any)`, because `TypeInformation.of(Any)` can only works for non-generic types. 

For example: 

```
public class UDAF extends AggregateFunction[Tuple2<String, Long>, Long] {
    public Long createAccumulator() {...}
    public Tuple2<String, Long> getValue(Long acc) {...}
}
```
For the given UDAF, the return type is a generic type `Tuple2<String, Long>`. We can't extract the return type  from an instance of Tuple2, because of Java type erasure. But we can extract it from the type hierarchy of UDAF class. 

Am I right, @twalthr ?



```

wuchong commented:
Comments addressed

wuchong commented:
I have rebased the code and updated.  Would you like to have a look again? @fhueske  @twalthr @sunjincheng121

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala:
```diff
@@ -320,12 +347,20 @@ object UserDefinedFunctionUtils {
     } else if(extractedType != null) {
```
+space `} else if (extractedType`

fhueske commented on flink-libraries/flink-table/src/test/java/org/apache/flink/table/api/java/utils/UserDefinedAggFunctions.java:
```diff
@@ -59,7 +59,7 @@ public boolean requiresOver() {
 	/**
 	 * Accumulator for WeightedAvg.
 	 */
-	public static class WeightedAvgAccum extends Tuple2<Long, Integer> {
```
Why do you change the type of this accumulator?

wuchong replied:
TypeExtractor will recognize `WeightedAvgAccum` as a `TupleType` which will ignore the `sum` and `count` fields when de/serialize. So I changed it as a POJO class. This should not be included in this PR. But I fixed it when I find it.

fhueske replied:
Yes, you are right. Either we change the type or we touch the code and use `acc.f0` instead of `acc.sum` and `acc.f1` instead of `acc.count`.

Let's go with your approach.

fhueske commented:
+1 to merge

wuchong commented:
merging...
